working/extension_structs/overview.md

```diff
@@ -0,0 +1,870 @@
+extensions to the core proposal are described in line.
+
+## Extension structs by example
+
+Extension structs are targeted at relatively niche uses where you wish to define
+a set of statically dispatched methods layered on top of an underlying
+representation, without introducing a wrapper object.  A canonical use case
+driving this design is to be able define a Dart typed interface for methods on a
+Javascript object, providing a wrapperless interoperation capability.
+
+
+```dart
+extension struct Window(JSObject o) {
+  // Signatures provided here for methods to be delegated to the underlying
+  // JSObject
+  external bool get closed;
+  // etc
+}
+
+void test(Window w) {
+  // Window methods can be called using Dart syntax
+  if (w.closed) {...}
+
+  // Windows are represented as the underlying object
+  assert(w is JSObject);
+
+  // Windows can be cast to the underlying object
+  JSObject o = w as JSObject;
+
+  // Extension struct types are reified as the underlying representation type.
+  List<Window> l = [w];
+  assert(l is List<JSObject>);
+}
+```
+
+Extension structs are also useful for providing a lightweight facade over an
+existing type.
+
+```dart
+// Natural numbers
+extension struct Nat(int _x) {
+  Nat(int x) : assert(x >= 0), _x = x;
+  Nat.zero() : _x = 0;
+
+  Nat get succ => Nat(_x+1);
+  Nat plus(Nat other) => Nat._x + other._x;
+
+  // Override the underlying isNegative operation (incorrectly)
+  bool get isNegative => true;
+}
+
+void test() {
+  var n1 = Nat(3);
+  var n2 = Nat.zero();
+  assert(n2.succ.succ.succ == n2);
+
+  //The underlying representation is still as an int
+  assert(n1 is int);
+
+  // The static type is used to dispatch the method calls
+  assert(n1.isNegative);
+
+  // If the static type is lost, dispatch goes to the underlying object.
+  dynamic d = n1;
+  assert(!d.isNegative);
+}
+```
+
+Extension structs may delegate members to the underlying field.  They may also
+implement interfaces, but only if the underlying field implements the interface.
+
+```dart
+// Natural numbers
+extension struct Nat(int _x) implements Comparable<num> {
+  // Constructors etc as above
+
+
+  bool get isEven;  // Abstract definition delegates to _x.isEven
+}
+
+void test() {
+  var n = Nat(2);
+  // Same as 2.isEven
+  assert(n.isEven);
+  // Comparable interface allows access to the Comparable methods on int
+  assert(n.compareTo(2) == 0);
+
+  // Since Nat implements Comparable<num>, it may be assigned to it
+  Comparable<num> c = n;
+  // The representation is still as an integer
+  assert(c is int);
+}
+```
+
+Extension structs have no inheritance.
+
+```dart
+// Static error
+extension struct PositiveNumber(int _x) extends Nat {...}
+```
+
+Extension structs do not define a signature, and hence cannot be implemented.
+
+```dart
+// Static error
+extension struct AlternativeNat(int _x) implements Nat { ...}
+
+// Static error
+class MockNat implements Nat {...}
+```
+
+
+## Extension structs in more detail
```

Comment:
I'm not really convinced that `extension structs` should be `structs`: The problem is that an extension struct may have an interface which is the entire interface of the underlying representation (the type of the unique instance variable), and then some added members, and there is no reason to assume that this is a proper interface for an immutable entity to have.

If we hadn't had this ability to skip the indirection (so we can do `myES.foo()` meaning `myES.uniqueField.foo()`) then the outermost object (existing or not) would be a proper (shallow-)immutable object, and everything would fit nicely. On the other hand, that object wouldn't be able to do _anything_ other than calling the getter `uniqueField` (plus, presumably the members of `Object`).

When we conflate the wrapped object with the wrapper (whether or not it exists), we get this conflict.

For instance, if we were to introduce some kind of boxing, I'd assume that the boxed class would simply emerge by taking the `extension struct` declaration and delete the word `extension`. We would then have a declaration that specifies a wrapping entity, wrapping a single object. It would be a regular, proper object, and it would have OO dispatch for its methods, and so on.

However, that's a struct, but it would now be a struct that contains forwarding methods to the representation object, including, possibly, some getters and setters which are forwarding to the implicitly induced getters and setters of the mutable instance variables that the representation object has.

This is a weird kind of `struct`, because it's simply a forwarding wrapper that exposes the entire interface of the wrapped object, including whatever mutability that object offers.

Why, then, would we even have the connection between the `struct` mechanism and the view mechanism known as `extension struct`? As far as I can see we could just as well use a normal class when we're boxing an extension struct.

Reply:
I'm open to splitting out extension structs and structs (e.g. into `data class` and `view class`), but I mostly think the mutability argument is a red-herring.  Structs aren't deeply immutable, and you can freely add forwarding methods that mutate the underlying fields.  Similarly, and extension struct is shallow (not deeply) immutable in the sense that it's not an lvalue, but it does allow delegating mutating members to the unique field.   It's true it makes it easier to do so in a way that somewhat obscures the difference though.

Put another way, any interface that can be used in the implements clause of an extension struct can also be used in the interface of a regular struct.